### PR TITLE
(Fix #541) Raise limit for ban/exempt/invite/ignore from 365 days to 2000 days

### LINKED ISSUE #541: .+ban durations are capped at 365 days
When wanting to set a ban like so, for a duration of 2 years, so 730 days:
`.+ban *!*@2001:db8::/32 #channel %730d You have been banned for... something.`

Then verifying it with:
`.bans *!*@2001:db8::/32`
```
Channel bans for #channel:  (* = not placed by bot)
! [  1] *!*@2001:db8::/32 (expires in 364 days)
        Robby: You have been banned for... something.
        Created 22:32
```
The duration seems to be capped at 365 days.

Not sure if this is intended or a bug (because why cap this?), but it would be great to have this cap removed, so longer bans can be set.

### PR DESCRIPTION
**Pull request by michaelortmann**
Found by: Robby
Patch by: michaelortmann
Fixes: #541

One-line summary:
old limit was arbitrary 365 days (~1 year) and workaroundable; new limit is 2000 days all specifiers combined.

Additional description (if needed):
old limit was arbitrary 365 days (~1 year) and workaroundable in 2 different ways (i like to exploit things :P):
1. if you did all 3 specifiers %d%h%m the combined limit was ~3 years.
2. specifiers are loosely combined and only checked one by one, so you could exploit the limit:
```
.+ignore hackmelimit %400d400d400d400d400d400d
Now ignoring: hackmelimit!*@* (requested)
[00:43:22] #-HQ# +ignore hackmelimit!*@* requested
.ignores
[00:43:26] #-HQ# ignores 
Currently ignoring:
  [  1] hackmelimit!*@* (expires in 2189 days)
        -HQ: requested
        Started 00:43
```
new limit could be max. ~19years.
so that time in seconds doesnt overflow (INT_MAX - now).
eggdrop suffers like most? software from year 2038 problem, see:
https://en.wikipedia.org/wiki/Year_2038_problem
this new limit would be dynamic.
it was decided to set a static combined limit of 2000 days.
this is what this patch does.
see the troll come back in 2000 days.
capped commands are:
.+ban
.+exempt
.+invite
.+ignore
Added new specifiers %y (1 year = 365 days) and %M (1 month = 30 days)

Test cases demonstrating functionality (if applicable):

```
.+ban test1 %2000d
[03:51:30] #-HQ# (GLOBAL) +ban test1!*@* (requested)
New ban: test1!*@* (requested)

.+ban test2 %2001d
Limit for expire time is 2000 days.

.+ban test3 %2000d1h
Limit for expire time is 2000 days.

.+ban test4 %2000d1m
Limit for expire time is 2000 days.

.bans all
[03:51:43] #-HQ# bans all
Global bans:
  [  1] test1!*@* (expires in 1999 days)
        -HQ: requested
        Created 03:51

.+exempt test5 %2000d
[03:51:48] #-HQ# (GLOBAL) +exempt test5!*@* (requested)
New exempt: test5!*@* (requested)

.+exempt test6 %2001d
Limit for expire time is 2000 days.

.exempts all
[03:51:56] #-HQ# exempts all
Global exempts:
  [  1] test5!*@* (expires in 1999 days)
        -HQ: requested
        Created 03:51

.+invite test7 %2000d
[03:52:01] #-HQ# (GLOBAL) +invite test7!*@* (requested)
New invite: test7!*@* (requested)

.+invite test8 %2001d
Limit for expire time is 2000 days.

.invites all
[03:52:09] #-HQ# invites all
Global invites:
  [  1] test7!*@* (expires in 1999 days)
        -HQ: requested
        Created 03:52

.+ignore test9 %2000d
Now ignoring: test9!*@* (requested)
[03:52:14] #-HQ# +ignore test9!*@* requested

.+ignore test10 %2001d
Limit for expire time is 2000 days.

.ignores    
[03:52:28] #-HQ# ignores 
Currently ignoring:
  [  1] test9!*@* (expires in 1999 days)
        -HQ: requested
        Started 03:52

.+ignore test11 %-1d 
expire time must be equal to or less than 2000 days

.+ignore test12 %1d-6h
Now ignoring: test12!*@* (requested)
[00:32:37] #-HQ# +ignore test12!*@* requested

.ignores 
Currently ignoring:
  [  1] test12!*@* (expires at 18:32)
        -HQ: requested
        Started 00:32
```

**Test new specifiers %y (1 year = 365 days) and %M (1 month = 30 days)**

```
.+ignore test13 %1y1M
Now ignoring: test13!*@* (requested)
[02:41:55] #-HQ# +ignore test13!*@* requested

.ignores
[02:41:58] tcl: builtin dcc call: *dcc:ignores -HQ 1 
[02:41:58] #-HQ# ignores 
Currently ignoring:
  [  1] test13!*@* (expires in 394 days)
        -HQ: requested
        Started 02:41

.+ban test14 %1y1M
[02:44:19] #-HQ# (GLOBAL) +ban test14!*@* (requested)
New ban: test14!*@* (requested)

.bans
[02:44:22] tcl: builtin dcc call: *dcc:bans -HQ 1 
[02:44:22] #-HQ# bans 
Global bans:
  [  1] test14!*@* (expires in 394 days)
        -HQ: requested
        Created 02:44

.+exempt test15 %1y1M
[02:45:02] #-HQ# (GLOBAL) +exempt test15!*@* (requested)
New exempt: test15!*@* (requested)

.exempts
[02:45:04] tcl: builtin dcc call: *dcc:exempts -HQ 1 
[02:45:04] #-HQ# exempts 
Global exempts:
  [  1] test15!*@* (expires in 394 days)
        -HQ: requested
        Created 02:45

.+invite test16 %1y1M
[02:45:48] #-HQ# (GLOBAL) +invite test16!*@* (requested)
New invite: test16!*@* (requested)

.invites
[02:45:51] tcl: builtin dcc call: *dcc:invites -HQ 1 
[02:45:51] #-HQ# invites 
Global invites:
  [  1] test16!*@* (expires in 394 days)
        -HQ: requested
        Created 02:45

.+ban testminute %42m
[02:46:50] #-HQ# (GLOBAL) +ban testminute!*@* (requested)
New ban: testminute!*@* (requested)

.bans    
[02:46:56] tcl: builtin dcc call: *dcc:bans -HQ 1 
[02:46:56] #-HQ# bans 
Global bans:
  [  1] testminute!*@* (expires at 03:28)
        -HQ: requested
        Created 02:46
```

Test successful.

### CODE REVIEW

**Comment by michaelortmann** [comment]
if you want history, see also #542

**Comment by ghost** [comment]
I see a couple of issues. The network is a mess right now, so I'll list them here instead.

I had two commends about the latest commit.

(1) all the calculations over and over: doesn't that eat some CPU? I mean, it's MINISCULE, but still: repeating the same thing over and over = greater chance for typoes? Should 60s * 60m = TIME_HOUR and 60s * 60m * 24h = TIME_DAY (eggdrop.h)? That way, those who don't know that 86400s = 1 day won't have to look it up / calculate it to run through the code?

(2) If you use a ban time of %195804h, then the complaint will say: "Please select a lower value than 2.26625" (if "%li will return that value). That's confusing because, while that calculation means DAYS, the user used HOURS, the complaint text is ambiguous. The user will think: I can only set a ban for less than 3 hours? (A) clarify "days" in the complain, or, (B) respond in the same format they used (if multiple, pick the highest ranked one: if h is used, reply in h; elseif m is used, calculate in m: else calculate in s).

**Comment by michaelortmann** [comment]
(1a) doesnt eat cpu. any compiler will precalc such things and hardcode the result only
(1b) it was hardcoded 86400s before, so i changed it into 60s * 60m * 24h to solve exactly the human read/undestand issue. the source is better now than it was.
(1c) yes, you could argue the duplication of code here is bad, but it was already such a duplication of much larger chunks. in my initial fix i also said it would be nice we could get rid of code duplication here. but it will be merges now as is, ive got other work to do for now. its hard i know, but at one point i must stop, else i wouldnt finish any pull request at all.
(2) i changed the log message to only say
expire time must be equal to or less than 2000 days\n
so it makes sense.
i guess while you were analyzing my patch i was doing some more commits the same time.

NOW you can reciew.
and before you complain about the error message, i oriented myself now on the .tcl rand command error message.

**Comment by Cizzle** [comment]
@vanosg is that patchlevel update still intended here?
LGTM though I see you moved some var declarations up.

**Comment by michaelortmann** [comment]
i did reset the version.h update with my pull request.

the var moving is fine. i prefer variable declarations in function header.

File: src/cmds.c [hunk 1]
```diff
@@ -2595,38 +2595,33 @@ static void cmd_pls_ignore(struct userrec *u, int idx, char *par)
...
-    char *p, *p_expire;
-    unsigned long int expire_foo;
-
     p = newsplit(&par);
     p_expire = p + 1;
     while (*(++p) != 0) {
       switch (tolower((unsigned) *p)) {
       case 'd':
         *p = 0;
         expire_foo = strtol(p_expire, NULL, 10);
```
Comment (thommey) [inline src/cmds.c:2610]: I realize it has always been this way, but we should probably use a signed long there, %-1d shouldn't be valid (and not wrap around)

**Comment by michaelortmann** [comment]
@thommey:

i made expire_foo long instead of unsigned long.

but i think negative values are not invalid but a feature. %1d-6h should be fine and equivalent to %18h.

i added 2 new test cases (test 11 and test12) to #577.

ready for review again.

**Comment by Cizzle** [comment]
LGTM

**Review by thommey — APPROVED** [review]

**Comment by michaelortmann** [comment]
**IRC discussion resulted in:**

We want to also add the following 2 specifiers.
%M == months (notice the capital M; m == minutes; yes, our specifiers are switched from what strftime() does; were borked again)
%y == years

1 year doesn't equal a fixed amount of days in real life, but if financial institutes workaround this by https://en.wikipedia.org/wiki/Day_count_convention then eggdrop can do so, also :)

We define 1year == 365 days and 1 month == 30 days.

**Comment by vanosg** [comment]
Different day, different discussion. Setting the max to 5 years instead of arbitrary round number

```
.+ban *!*@foo.com %5y
New ban: *!*@foo.com (requested)

.+ban *!*@foo.com %5y1m
Ban expiration time cannot exceed 5 years (1825 days)

.+ban *!*@goo.com %1y1d1h1m
New ban: *!*@goo.com (requested)

.+ban *!*@doo.com
New ban: *!*@doo.com (requested)

.+ban *!*@woo.com %-1y foo
Ban expiration time cannot exceed 5 years (1825 days)
```

**Pull request merged**